traversal/walk.go

```diff
@@ -203,6 +203,9 @@ func (prog Progress) walkAdv(n datamodel.Node, s selector.Selector, fn AdvVisitF
 	if attn == nil {
 		return prog.walkAdv_iterateAll(n, s, fn)
 	}
+	if len(attn) == 0 {
+		return nil
+	}
```

Comment:
personally, I don't think treating a nil slice different from an empty slice is a good idea. when consuming those values, they should be treated as identical, just like with nil vs empty maps.

the standard library does this, too; I don't think there's a single API that treats a nil slice differently from an empty slice. it will even sometimes turn one into the other, because they are considered identical. see for instance https://github.com/golang/go/issues/46415#issuecomment-849953262.

if you want the `Interests` method to signal something here, then we need another value, I think.

Reply:
> I also added a shortcut to traversals so that a selector that explicitly states it's not interested in anything (by returning a non-nil but empty slice for its interests) is treated differently than a nil return for interests (which means "I don't know; hit me with everything you've got and lemme see").

The commit message is turning my mild concern into a somewhat severe one, given that the two values that should be treated identically now mean very different things as well :)

Reply:
I sort of agree, but how else would we like to do this?

Refactor to `Interests() (bool, []datamodel.PathSegment)`?

I guess that's certainly within reach; we do control all the implementations, and I don't believe this is a functionally public surface area, even if it technically is (I don't think any downstreams touch this).

Reply:
If you're fairly certain this is a closed interface that downstreams shouldn't use, then I agree we can just break it. We're a v0, after all, so we might as well make use of its advantages when necessary.

An alternative is to deprecate `Interests` in favor of another method with the new signature, and type-assert whether the interface has the new method or not. Uglier for sure, but more backwards compatible.

Reply:
It's definitely not used by downstreams.  We could change the signature.

I think I'm going to skip that change for now, though.  On re-read, the documentation on the `Interests` method already phrased this fairly carefully.  Changing the code further feels like increasing the size of the diff without much real impact.

I've added another sentence of clarifying doc, though.